src/Sema.zig

```diff
@@ -32281,6 +32281,46 @@ fn analyzeSlice(
                 break :e try sema.coerce(block, .usize, uncasted_end, end_src);
             } else break :e try sema.coerce(block, .usize, uncasted_end_opt, end_src);
         }
+
+        // when slicing a many-item pointer, if a sentinel `S` is provided as in `ptr[a.. :S]`, it
+        // must match the sentinel of `@TypeOf(ptr)`.
```

Comment:
(No action needed from this comment, just thinking out loud)

I don't think this rule actually makes sense; after all, `a` could be past the sentinel, meaning that `ptr[a.. :s]` can exhibit uncheckable Illegal Behavior. However, this aligns with the fact that `ptr[a..]` on a sentinel-terminated many-pointer currently preserves the sentinel, so changing this can be a separate proposal.